core/templates/pages/topic-editor-page/services/topic-editor-staleness-detection.service.ts

```diff
@@ -0,0 +1,124 @@
+import { EntityEditorBrowserTabsInfo } from 'domain/entity_editor_browser_tabs_info/entity-editor-browser-tabs-info.model';
+import { FaviconService } from 'services/favicon.service';
+import { LocalStorageService } from 'services/local-storage.service';
+import { UndoRedoService } from 'domain/editor/undo_redo/undo-redo.service';
+import { StaleTabInfoModalComponent } from 'components/stale-tab-info/stale-tab-info-modal.component';
+import { UnsavedChangesStatusInfoModalComponent } from 'components/unsaved-changes-status-info/unsaved-changes-status-info-modal.component';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TopicEditorStalenessDetectionService {
+  _staleTabEventEmitter = new EventEmitter<void>();
+  _presenceOfUnsavedChangesEventEmitter = new EventEmitter<void>();
+  unsavedChangesWarningModalRef: NgbModalRef = null;
+
+  constructor(
+    private ngbModal: NgbModal,
+    private windowRef: WindowRef,
+    private stalenessDetectionService: StalenessDetectionService,
+    private topicEditorStateService: TopicEditorStateService,
+    private faviconService: FaviconService,
+    private localStorageService: LocalStorageService,
+    private undoRedoService: UndoRedoService
+  ) {}
+
+  init(): void {
+    this.staleTabEventEmitter.subscribe(() => {
+      this.showStaleTabInfoModal();
+    });
+    this.presenceOfUnsavedChangesEventEmitter.subscribe(() => {
+      this.showPresenceOfUnsavedChangesModal();
+    });
+    this.topicEditorStateService.onTopicInitialized.subscribe(() => {
+      this.showStaleTabInfoModal();
+      this.showPresenceOfUnsavedChangesModal();
+    });
+  }
+
+  showStaleTabInfoModal(): void {
+    const topic = this.topicEditorStateService.getTopic();
+    const topicEditorBrowserTabsInfo: EntityEditorBrowserTabsInfo = (
+      this.localStorageService.getEntityEditorBrowserTabsInfo(
+        EntityEditorBrowserTabsInfoDomainConstants
+          .OPENED_TOPIC_EDITOR_BROWSER_TABS, topic.getId()));
+
+    if (
+      topicEditorBrowserTabsInfo &&
+      topicEditorBrowserTabsInfo.getLatestVersion() !== topic.getVersion()
+    ) {
+      this.faviconService.setFavicon(
+        '/assets/images/favicon_alert/favicon_alert.ico');
+      this.ngbModal.dismissAll();
+      const modalRef = this.ngbModal.open(
+        StaleTabInfoModalComponent, {
+          backdrop: 'static',
+        });
+      modalRef.componentInstance.entity = 'topic';
+      modalRef.result.then(() => {
+        this.windowRef.nativeWindow.location.reload();
+      }, () => {});
+    }
+  }
+
+  showPresenceOfUnsavedChangesModal(): void {
+    if (this.undoRedoService.getChangeCount() === 0) {
+      if (
+        this.stalenessDetectionService
+          .doesSomeOtherEntityEditorPageHaveUnsavedChanges(
+            EntityEditorBrowserTabsInfoDomainConstants
+              .OPENED_TOPIC_EDITOR_BROWSER_TABS,
+            this.topicEditorStateService.getTopic().getId())
+      ) {
```

Comment:
can we combine these if conditions?

Reply:
Yes we can but this might decrease readability of the code. What about the following:

```
    if (this.undoRedoService.getChangeCount() !== 0) {
      return;
    }
    if (
      this.stalenessDetectionService
        .doesSomeOtherEntityEditorPageHaveUnsavedChanges(
          EntityEditorBrowserTabsInfoDomainConstants
            .OPENED_TOPIC_EDITOR_BROWSER_TABS,
          this.topicEditorStateService.getTopic().getId())
    ) {
      this.ngbModal.dismissAll();
      this.unsavedChangesWarningModalRef = this.ngbModal.open(
        UnsavedChangesStatusInfoModalComponent, {
          backdrop: 'static',
        });
      this.unsavedChangesWarningModalRef.componentInstance.entity = 'topic';
      this.unsavedChangesWarningModalRef.result.then(() => {}, () => {});
    } else if (this.unsavedChangesWarningModalRef) {
      this.unsavedChangesWarningModalRef.dismiss();
    }
```

Reply:
LMK if this looks good.

Reply:
Yup this is fine.

Reply:
Please fix this in the other editor related PRs as well. Thanks!